### PR TITLE
[REBASED TO DEATH] Implement show and toggle as standard actions

### PR DESCRIPTION
**Pull request by alanorozco**

### CODE REVIEW

File: src/service/standard-actions-impl.js [hunk 1]
```diff
@@ -96,6 +99,34 @@ export class StandardActions {
...
+      }
+    });
+  }
+
+  /**
+   * Handles "toggle" action.
+   * @param {!./action-impl.ActionInvocation} invocation
+   */
+  handleToggle(invocation) {
+    if (getStyle(dev().assertElement(invocation.target), 'display') == 'none') {
```
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: we should use `getComputed` style here so if element is hidden some other way (e.g. author defined CSS class) we toggle properly.

File: test/functional/test-standard-actions.js [hunk 2]
```diff
@@ -23,36 +23,109 @@ import {setParentWindow} from '../../src/service';
...
+
+      standardActions.handleShow({target: element});
+      expect(deferMutateStub).to.be.calledOnce;
+      expect(deferMutateStub.firstCall.args[0]).to.equal(element);
+      expect(element.expand).to.be.calledOnce;
+    });
+
+  });
+
+  describe('"toggle" action', () => {
```
Comment (aghassemi) [inline test/functional/test-standard-actions.js]: and a test here to handle the comment above about computedStyle. Something along the lines of creating a `style` tag with a class like `.hide { display: none }` and using that class instead of `display=none`.

File: test/functional/test-standard-actions.js [hunk 3]
```diff
@@ -23,36 +23,109 @@ import {setParentWindow} from '../../src/service';
...
     });
 
     it('should handle AmpElement', () => {
+      const element = createAmpElement();
+      element.expand = sandbox.stub();
+
+      standardActions.handleShow({target: element});
+      expect(deferMutateStub).to.be.calledOnce;
+      expect(deferMutateStub.firstCall.args[0]).to.equal(element);
+      expect(element.expand).to.be.calledOnce;
```
Comment (aghassemi) [inline test/functional/test-standard-actions.js]: also `expectElementToHaveBeenShown(element);` here?

File: test/functional/test-standard-actions.js [hunk 4]
```diff
@@ -23,36 +23,109 @@ import {setParentWindow} from '../../src/service';
...
+
+  describe('"show" action', () => {
+    it('should handle normal element', () => {
+      const element = document.createElement('div');
+      standardActions.handleShow({target: element});
+      expectElementToHaveBeenShown(element);
     });
 
     it('should handle AmpElement', () => {
+      const element = createAmpElement();
```
Comment (aghassemi) [inline test/functional/test-standard-actions.js]: Let's also add a test where an AMP document is hidden using `layout=nodisplay` and then a show action is called on it ensuring `display:none` added by `layout=nodisplay` goes away after the action.

File: src/service/standard-actions-impl.js (hunk 1 above)
Comment (alanorozco) [inline src/service/standard-actions-impl.js]: Done. Using vsync to avoid layout thrashing.

File: test/functional/test-standard-actions.js (hunk 2 above)
Comment (alanorozco) [inline test/functional/test-standard-actions.js]: Done. Stubbed vsync and added elements to the body tag in the test so that `getComputedStyle` can work properly.

File: test/functional/test-standard-actions.js (hunk 4 above)
Comment (aghassemi) [inline test/functional/test-standard-actions.js]: true, nodisplay is equal to elem.style.display=none test case

**Comment by aghassemi** [comment]
@alanorozco looks great. One last request, could you please add a new file to `examples` like `global-actions.amp.html` that has bare-bone example usage for `hide`, `show`, `toggle` (ideally for both html and amp elements as targets)

**Comment by aghassemi** [comment]
@alanorozco FYI (RE:getComputedStyle): https://github.com/ampproject/amphtml/pull/7733

**Comment by alanorozco** [comment]
Standing by before decision on [CSS-defined property issue](https://github.com/ampproject/amphtml/issues/7027#issuecomment-281847363) is made.

File: src/base-element.js [hunk 5]
```diff
@@ -862,6 +862,24 @@ export class BaseElement {
...
+   * is no longer visible.
+   */
+  expand() {
+    this.element.getResources().expandElement(this.element);
+  }
+
+  /**
+   * Called every time an owned AmpElement expands itself.
+   * See {@link expand}.
+   * @param {!AmpElement} unusedElement
```
Comment (dvoytenko) [inline src/base-element.js]: What is this argument actually? The element itself? If so, we already have `this.element` so it's not needed.

File: src/custom-element.js [hunk 6]
```diff
@@ -1323,6 +1323,24 @@ function createBaseCustomElementClass(win) {
     }
 
     /**
+     * Expands the element, and notifies its owner (if there is one) that the
+     * element is now present.
+     * @suppress {missingProperties}
```
Comment (dvoytenko) [inline src/custom-element.js]: What's the suppress for?

File: src/service/standard-actions-impl.js [hunk 7]
```diff
@@ -87,15 +100,57 @@ export class StandardActions {
...
+  /**
+   * Handles "show" action. This is a very simple action where "display: none"
+   * is removed from the target element.
+   * @param {!./action-impl.ActionInvocation} invocation
+   */
+  handleShow(invocation) {
+    // TODO(alanorozco, #7753) use 'hidden' attribute for AMP elements
+    const target = dev().assertElement(invocation.target);
+
+    user().assert(isHidden(target),
```
Comment (dvoytenko) [inline src/service/standard-actions-impl.js]: Let's remove this check. It's totally fine to ask a visible element to be shown again. We can just skip work if not needed.

File: spec/amp-actions-and-events.md [hunk 8]
```diff
@@ -61,7 +61,9 @@ Example: `on="submit-success:lightbox1;submit-error:lightbox2"`
 ## Globally defined Events and Actions
 Currently AMP defines `tap` event globally that you can listen to on any HTML element (including amp-elements).
 
-AMP also defines a `hide` action globally that you can trigger on any HTML element.
+AMP also defines `hide`, `show` and `toggle` actions globally that you can trigger on any HTML element.
+
+**Note:** An AMP element can only be shown if it was previously hidden by a `hide` or `toggle` action. Normal elements can set the `hidden` attribute if the element has to be hidden by default.
```
Comment (aghassemi) [inline spec/amp-actions-and-events.md]: `{% call callout('Note', type='note') %}Note: {% endcall %}`

Comment (aghassemi) [inline spec/amp-actions-and-events.md]: I would reword as:
An element can only be shown if it was previously hidden by a `hide` or `toggle` action, or by using the [hidden](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/hidden) attribute. `show` does **not** support elements hidden by CSS `display:none` or AMP's `layout=nodisplay`

@bpaduch for review.

File: src/service/standard-actions-impl.js [hunk 9]
```diff
@@ -87,15 +100,57 @@ export class StandardActions {
...
   handleHide(invocation) {
+    // TODO(alanorozco, #7753) use 'hidden' attribute for AMP elements
     const target = dev().assertElement(invocation.target);
+
     this.resources_.mutateElement(target, () => {
       if (target.classList.contains('i-amphtml-element')) {
         target./*OK*/collapse();
       } else {
         toggle(target, false);
+        target.setAttribute('hidden', '');
```
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: `completeCollapse` needs to do the same with the `hidden` attribute.

File: src/service/standard-actions-impl.js [hunk 10]
```diff
@@ -87,15 +100,57 @@ export class StandardActions {
...
+
+    // deferMutate will only work on AMP elements
+    if (target.classList.contains('i-amphtml-element')) {
+      this.resources_.deferMutate(target, () => {
+        target./*OK*/expand();
+      });
+    } else {
+      this.resources_.mutateElement(target, () => {
+        toggle(target, true);
+        target.removeAttribute('hidden');
```
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: `completeExpand` needs to do the same with 'hidden' attribute.

File: src/service/standard-actions-impl.js [hunk 11]
```diff
@@ -87,15 +100,57 @@ export class StandardActions {
...
+   * Handles "show" action. This is a very simple action where "display: none"
+   * is removed from the target element.
+   * @param {!./action-impl.ActionInvocation} invocation
+   */
+  handleShow(invocation) {
+    // TODO(alanorozco, #7753) use 'hidden' attribute for AMP elements
+    const target = dev().assertElement(invocation.target);
+
+    user().assert(isHidden(target),
+        'Element can only be shown when it has the "hidden" attribute set or ' +
```
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: `layout=nodisplay` will pass the `isHidden()` test. Maybe `isHidden(target) && layout != nodisplay`)

File: src/service/standard-actions-impl.js (hunk 7 above)
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: let's just do this in development mode and use `dev()` since calling `show()` on something that is already shown will throw and that's just too much.
```
if(getMode().development) {
  dev().assert(...)
}
```

Comment (aghassemi) [inline src/service/standard-actions-impl.js]: thinking about this a bit more, this has too much false positives. The purpose of this is to let the developer know about the cases where something is hidden but we can't show it so maybe something along the lines of:

```
isHidden = computedStyle == none || hasAttribute(hidden)
isShowable = (style == none && layout != nodisplay) || hasAttribute(hidden)
if( isHidden && !isShowable) {
 // warn the developer
}
```

Comment (alanorozco) [inline src/service/standard-actions-impl.js]: Actually there's no need to check for computedStyle since the inline style will take priority in its result.

File: src/service/standard-actions-impl.js (hunk 11 above)
Comment (alanorozco) [inline src/service/standard-actions-impl.js]: Done.

File: examples/standard-actions.amp.html [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  <h3>Normal element:</h3>
+  <button on="tap:normal-element.show">Show</button>
+  <button on="tap:normal-element.hide">Hide</button>
+  <button on="tap:normal-element.toggle">Toggle</button>
+</div>
+
+<amp-img src="https://ampbyexample.com/img/amp.jpg" layout="responsive" width="1080" height="610" id="img-on-viewport"></amp-img>
+
+<div id="normal-element" hidden>
+  Hide me if you like.
```
Comment (alanorozco) [inline examples/standard-actions.amp.html]: Done.

File: src/service/standard-actions-impl.js [hunk 13]
```diff
@@ -88,14 +103,62 @@ export class StandardActions {
    */
   handleHide(invocation) {
     const target = dev().assertElement(invocation.target);
+
     this.resources_.mutateElement(target, () => {
       if (target.classList.contains('i-amphtml-element')) {
         target./*OK*/collapse();
       } else {
         toggle(target, false);
+        target.setAttribute('hidden', '');
```
Comment (dvoytenko) [inline src/service/standard-actions-impl.js]: To confirm: do we need `hidden` attribute polyfill?

File: src/base-element.js (hunk 5 above)
Comment (dvoytenko) [inline src/base-element.js]: Ok. Let's document what this parameter means then in jsdoc.

File: src/custom-element.js (hunk 6 above)
Comment (dvoytenko) [inline src/custom-element.js]: Let's just try to remove and run lint/check-types and see if anyone complains.

File: src/service/standard-actions-impl.js (hunk 7 above)
Comment (dvoytenko) [inline src/service/standard-actions-impl.js]: Please describe it generously in comments. Further, is this an `error` or a `warn`? Errors are sent to our error monitoring. Do we need it?

File: src/service/standard-actions-impl.js (hunk 13 above)
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: I don't think so. Support has been around for a while http://caniuse.com/#feat=hidden All versions of Chrome and Firefox, iOS > 5, IE11, and all versions of Edge support it.

File: src/service/standard-actions-impl.js [hunk 14]
```diff
@@ -88,14 +103,62 @@ export class StandardActions {
...
+
+  /**
+   * Handles "show" action. This is a very simple action where "display: none"
+   * is removed from the target element.
+   * @param {!./action-impl.ActionInvocation} invocation
+   */
+  handleShow(invocation) {
+    const target = dev().assertElement(invocation.target);
+
+    if (getMode().development) {
```
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: Sorry, actually removing `getMode().development` check and using the `user().warn` is the right approach.  `getMode().development` won't help external developers. make sure you `return` after `user().warn`, unlike `assert` it does not throw synchronously to stop execution.

File: src/service/standard-actions-impl.js [hunk 15]
```diff
@@ -88,14 +103,62 @@ export class StandardActions {
...
+          'Elements with layout=nodisplay cannot be dynamically shown. %s',
+          target);
+
+      dev().assert(
+          target.style.display == 'none' || target.hasAttribute('hidden'),
+          'Elements can only be dynamically shown when they have the "hidden"' +
+          ' attribute set or when they were previously dynamically hidden. %s',
+          target);
+    }
+
```
Comment (aghassemi) [inline src/service/standard-actions-impl.js]: So the case still not covered for the warning (hence my comment regarding computedStyle) is if element is hidden using a css class rather than inline-style, `target.style.display` will be empty. So doing a similar check to above using computedStyle in a vsync seems needed for completeness.

File: src/service/standard-actions-impl.js (hunk 13 above)
Comment (dvoytenko) [inline src/service/standard-actions-impl.js]: We still support IE10 though, right? And I think we cut off at Chrome 37. Perhaps we can add it just in case to `amp.css`? I.e. 
```
[hidden] {
  display: none !important;
}
```
?

File: src/base-element.js (hunk 5 above)
Comment (alanorozco) [inline src/base-element.js]: Done.

File: src/custom-element.js (hunk 6 above)
Comment (alanorozco) [inline src/custom-element.js]: Done.

File: src/service/standard-actions-impl.js (hunk 13 above)
Comment (alanorozco) [inline src/service/standard-actions-impl.js]: Done.

File: src/service/standard-actions-impl.js (hunk 14 above)
Comment (alanorozco) [inline src/service/standard-actions-impl.js]: Done.

File: src/service/standard-actions-impl.js (hunk 15 above)
Comment (alanorozco) [inline src/service/standard-actions-impl.js]: True.

**Comment by alanorozco** [comment]
Closing due to rebase issue.

New PR is #7785
